### PR TITLE
adds 'SSL client certificate' support from n.octeau with qknight changes...

### PR DESCRIPTION
**Pull request by qknight**
... as:

* removed broken QSsl::SslV3 default
* rewrote slotHandleErrors(): no longer claim errors which are none.
* hack reverted: lib64 was not the cause for NixOS issues related to libraries.
* refactored csync/src/csync_owncloud.c and discovered+fixed why the dav_connect was never getting the certPath+certPassoword
* cleanup of code but seems this crushed the ssl client certificate support
* fixes the https://github.com/owncloud/client/issues/69#issuecomment-69358377 issue
* lots of cleanup
* From TODO list : translate all french comments into english
* changed _pemCertificate type from QString to QByteArray

### CODE REVIEW

File: csync/src/csync_owncloud.c [hunk 1]
```diff
@@ -826,12 +856,28 @@ int owncloud_set_property(CSYNC* ctx, const char *key, void *data) {
...
         if (data) {
             csync_owncloud_redirect_callback_t* cb_wrapper = data;
-
             ctx->owncloud_context->dav_session.redir_callback = *cb_wrapper;
         } else {
             ctx->owncloud_context->dav_session.redir_callback = NULL;
         }
     }
+    if( c_streq(key, "SSLClientCerts")) {
+        if(ctx->clientCerts != NULL) {
```
Comment (ogoffart) [inline csync/src/csync_owncloud.c]: You need to do the same cleanup in owncloud_commit

File: src/gui/CMakeLists.txt [hunk 2]
```diff
@@ -253,7 +257,7 @@ install(TARGETS ${APPLICATION_EXECUTABLE}
         BUNDLE  DESTINATION "."
  )
 
-# FIXME: The following lines are dup in src/gui and src/cmd because it needs to be done after both are installed
```
Comment (ogoffart) [inline src/gui/CMakeLists.txt]: Why did you remove that line?

File: src/gui/addcertificatedialog.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+class AddCertificateDialog : public QDialog
+{
+    Q_OBJECT
+
+public:
+    explicit AddCertificateDialog(QWidget *parent = 0);
+    ~AddCertificateDialog();
+    QString getCertificatePath();
+    QString getCertificatePasswd();
+    void showErrorMessage(QString message);
```
Comment (ogoffart) [inline src/gui/addcertificatedialog.h]: should be const QString &message  (pass the argument by const reference)

File: src/libsync/account.h [hunk 4]
```diff
@@ -144,6 +146,8 @@ class OWNCLOUDSYNC_EXPORT Account : public QObject {
     QVariant credentialSetting(const QString& key) const;
     void setCredentialSetting(const QString& key, const QVariant &value);
 
+    void setCertificate(QByteArray certficate = QByteArray(), QString privateKey = QString());
```
Comment (ogoffart) [inline src/libsync/account.h]: const QByteArray&  and const QString&

File: src/libsync/configfile.cpp [hunk 5]
```diff
@@ -255,11 +260,15 @@ QString ConfigFile::excludeFileFromSystem()
...
                 QLatin1String("../Resources/") + exclFile );
 #endif
+    
     return fi.absoluteFilePath();
 }
 
 QString ConfigFile::configFile() const
 {
+    if( qApp->applicationName().isEmpty() ) {
+        qApp->setApplicationName( Theme::instance()->appNameGUI() );
```
Comment (ogoffart) [inline src/libsync/configfile.cpp]: Can you explain why you need to do this?

File: src/libsync/configfile.cpp [hunk 6]
```diff
@@ -213,6 +216,8 @@ QString ConfigFile::excludeFile(Scope scope) const
     // prefer sync-exclude.lst, but if it does not exist, check for
     // exclude.lst for compatibility reasons in the user writeable
     // directories.
+    const QString exclFile("sync-exclude.lst");
```
Comment (ogoffart) [inline src/libsync/configfile.cpp]: i believe this is a merge error

**Comment by danimo** [comment]
Does not link on OS X. Testing windows now:

```
Linking CXX shared library libowncloudsync.dylib
Undefined symbols for architecture x86_64:
  "_BIO_ctrl", referenced from:
      x509ToString(bio_st*) in p12topem.cpp.o
  "_BIO_free", referenced from:
      p12ToPem(std::string, std::string) in p12topem.cpp.o
  "_BIO_new", referenced from:
      p12ToPem(std::string, std::string) in p12topem.cpp.o
  "_BIO_read", referenced from:
      x509ToString(bio_st*) in p12topem.cpp.o
  "_BIO_s_mem", referenced from:
      p12ToPem(std::string, std::string) in p12topem.cpp.o
  "_ERR_load_crypto_strings", referenced from:
      p12ToPem(std::string, std::string) in p12topem.cpp.o
  "_OPENSSL_add_all_algorithms_noconf", referenced from:
      p12ToPem(std::string, std::string) in p12topem.cpp.o
  "_PEM_write_bio_PrivateKey", referenced from:
      p12ToPem(std::string, std::string) in p12topem.cpp.o
  "_PEM_write_bio_X509", referenced from:
      p12ToPem(std::string, std::string) in p12topem.cpp.o
  "_PKCS12_free", referenced from:
      p12ToPem(std::string, std::string) in p12topem.cpp.o
  "_PKCS12_parse", referenced from:
      p12ToPem(std::string, std::string) in p12topem.cpp.o
  "_d2i_PKCS12_fp", referenced from:
      p12ToPem(std::string, std::string) in p12topem.cpp.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/libsync/libowncloudsync.1.8.0.dylib] Error 1
make[1]: *** [src/libsync/CMakeFiles/owncloudsync.dir/all] Error 2
make: *** [all] Error 2
```

File: src/3rdparty/certificates/p12topem.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <string>
+#include <openssl/pem.h>
+#include <openssl/err.h>
+#include <openssl/pkcs12.h>
+
+using namespace std;
```
Comment (ogoffart) [inline src/3rdparty/certificates/p12topem.h]: It would be better not to have using namespace in a header.

File: src/gui/CMakeLists.txt (hunk 2 above)
Comment (qknight) [inline src/gui/CMakeLists.txt]: by accident, will revert it

File: csync/src/csync_owncloud.c (hunk 1 above)
Comment (qknight) [inline csync/src/csync_owncloud.c]: i don't understand where i shoud put that cleanup code.

i tried to put it at the very end of this function but then the client receives a 'ssl handshacke error'

please provide details.

File: src/gui/addcertificatedialog.h (hunk 3 above)
Comment (qknight) [inline src/gui/addcertificatedialog.h]: changed

File: src/libsync/account.h (hunk 4 above)
Comment (qknight) [inline src/libsync/account.h]: changed

File: src/libsync/configfile.cpp (hunk 5 above)
Comment (qknight) [inline src/libsync/configfile.cpp]: merge error -> changed!

File: src/libsync/configfile.cpp (hunk 6 above)
Comment (qknight) [inline src/libsync/configfile.cpp]: correct!

File: src/3rdparty/certificates/p12topem.h (hunk 7 above)
Comment (qknight) [inline src/3rdparty/certificates/p12topem.h]: removed it

File: csync/src/csync_owncloud.c (hunk 1 above)
Comment (ogoffart) [inline csync/src/csync_owncloud.c]: Normaly it should be in owncloud_commit
Or maybe in owncloud_destruct.

**Comment by danimo** [comment]
Will merge and fix platform issues in a separate commit

**Comment by danimo** [comment]
Hmm, doesn't merge cleanly any longer, so I will perform a manual merge

**Comment by ogoffart** [comment]
merged
